### PR TITLE
mgr/progress: improve+test OSD out handling

### PR DESCRIPTION
**Pull request by jcsp**
This is a follow-up to flesh out the part of the progress module that generates progress events on OSDs going out.

### CODE REVIEW

**Comment by jcsp** [comment]
Tests passing here: http://pulpito.ceph.com/jspray-2018-07-30_14:30:34-rados:mgr-wip-progress-distro-basic-smithi/

**Comment by jcsp** [comment]
retest this please

File: src/pybind/mgr/progress/module.py [hunk 1]
```diff
@@ -409,18 +453,30 @@ def _handle_ls(self):
...
+            'events': [ev.to_json() for ev in self._events.values()],
+            'completed': [ev.to_json() for ev in self._completed_events]
+        }
+
     def _handle_clear(self):
         self._events = {}
+        self._completed_events = []
+        self._dirty = True
+        self._save()
+
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py]: might want to clear `self._dirty` here? just to be consistent with `self.serve()` ? or better off, just rename `_save()` to `_maybe_save()`, and do the "check the flag, encode, set_store, clear the flag" dance there?

File: src/pybind/mgr/progress/module.py [hunk 2]
```diff
@@ -130,6 +130,10 @@ def __init__(self, message, refs, which_pgs, evactuate_osds):
 
         self.id = str(uuid.uuid4())
 
+    @property
+    def evacuating_osds(self):
+        return self. _evacuate_osds
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/progress/module.py]: or we can just add a method named `on_osd_in(osd_id)` to `Event` and let the subclass of `Event` to decide if it should complete at seeing this message, i.e., `self._progress = 1.0`.

File: src/pybind/mgr/progress/module.py [hunk 3]
```diff
@@ -260,17 +259,49 @@ def __init__(self, *args, **kwargs):
...
                     "up_acting: {0}".format(json.dumps(up_acting, indent=2)))
-                if osd_id in up_acting['up'] or osd_id in up_acting['acting']:
+
+                new_up_acting = new_map.pg_to_up_acting_osds(pool['pool'], ps)
+                new_osds = set(new_up_acting['up']) | set(new_up_acting['acting'])
+
+                # Has this OSD been assigned a new location?
+                # (it might not be if there is no suitable place to move
+                #  after an OSD failure)
+                is_relocated = len(new_osds - old_osds) > 0
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py]: instead, might want to put
```py
is_relocated = osd_id in old_osds - new_osds
```

File: src/pybind/mgr/progress/module.py [hunk 4]
```diff
@@ -260,17 +259,49 @@ def __init__(self, *args, **kwargs):
...
+
+                self.log.debug(
+                    "new_up_acting: {0}".format(json.dumps(new_up_acting,
+                                                           indent=2)))
+
+                if was_on_out_osd and is_relocated:
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py]: `was_on_out_osd` is always `True` here.

File: src/pybind/mgr/progress/module.py [hunk 5]
```diff
@@ -260,17 +259,49 @@ def __init__(self, *args, **kwargs):
...
+        unmoved_pgs = []
         for pool in old_dump['pools']:
             pool_id = pool['pool']
             for ps in range(0, pool['pg_num']):
+                up_acting = old_map.pg_to_up_acting_osds(pool['pool'], ps)
+
+                # Was this OSD affected by the OSD going out?
+                old_osds = set(up_acting['up']) | set(up_acting['acting'])
+                was_on_out_osd = osd_id in old_osds
+                if not was_on_out_osd:
```
Comment (tchaikov) [inline src/pybind/mgr/progress/module.py]: no need to assign it to a variable, could just put
```py
if osd_id not in old_osds:
  continue
```

File: src/pybind/mgr/progress/module.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: space removed.  I think we might want to adopt a style like on_osd_in (+ more generally hooks for map updates on events) but that's probably a refactor for later.

File: src/pybind/mgr/progress/module.py [hunk 6]
```diff
@@ -269,6 +289,14 @@ def _osd_out(self, old_map, old_dump, new_map, osd_id):
         ev.pg_update(self.get("pg_dump"), self.log)
         self._events[ev.id] = ev
 
+    def _osd_in(self, osd_id):
+        for ev_id, ev in self._events.items():
+            if isinstance(ev, PgRecoveryEvent) and osd_id in ev.evacuating_osds:
+                self.log.info("OSD {0} came back in, cancelling event".format(
```
[the remark this answers is not in the document]
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: I somehow forgot all about that more efficient style -- changed this now

File: src/pybind/mgr/progress/module.py [hunk 7]
```diff
@@ -301,23 +335,25 @@ def notify(self, notify_type, notify_data):
             for ev_id, ev in self._events.items():
                 if isinstance(ev, PgRecoveryEvent):
                     ev.pg_update(data, self.log)
-        else:
-            self.log.debug("Ignore notification '{0}'".format(notify_type))
-            pass
+                    self.maybe_complete(ev)
 
-    def _persist(self):
+    def maybe_complete(self, event):
```
[the remark this answers is not in the document]
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: nope, underscore added

File: src/pybind/mgr/progress/module.py (hunk 1 above)
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: Yep, done.

File: src/pybind/mgr/progress/module.py (hunk 3 above)
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: I'm not sure I get the proposed logic.  At this point in the code, it's always true that that old_osds contains osd_id (that was checked a few lines earlier).  Given that, the check you're suggesting is really equivalent to "osd_id not in new_osds".

My intent is to only create the progress event when there are some new OSDs, i.e. there is some data to be moved.

File: src/pybind/mgr/progress/module.py (hunk 4 above)
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: Fixed.

File: src/pybind/mgr/progress/module.py (hunk 5 above)
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: done.

File: src/pybind/mgr/progress/module.py (hunk 3 above)
Comment (tchaikov) [reply src/pybind/mgr/progress/module.py]: got it! i misunderstood your intent, thought that the progress event was emitted only if the `osd_id` was *removed* from the `old_osds`.

**Comment by tchaikov** [comment]
@jcsp did you forget to update this PR?

**Comment by jcsp** [comment]
Oops, forgot to push this while retesting locally.  Updated now (+this passes test_progress on a vstart cluster)

File: src/pybind/mgr/progress/module.py [hunk 8]
```diff
@@ -167,20 +177,27 @@ def pg_update(self, pg_dump, log):
...
+
+                        # Since the recovered bytes (over time) could perhaps
+                        # exceed the contents of the PG (moment in time), we
+                        # must clamp this
+                        ratio = min(ratio, 1.0)
+
+                    else:
+                        # Dataless PGs (e.g. containing only OMAPs) count
+                        # as half done.
+                        ratio = 0.5
```
Comment (noahdesu) [inline src/pybind/mgr/progress/module.py]: If bytes aren't (currently) available as a metric for omap data, would it be possible to substitute a separate metric like the number of objects recovered? Are there any applications that store a significant amount of data in omap?

File: src/pybind/mgr/progress/module.py [hunk 9]
```diff
@@ -301,23 +335,25 @@ def notify(self, notify_type, notify_data):
...
+            self._complete(event)
+
+    def _save(self):
         self.log.info("Writing back {0} completed events".format(
             len(self._completed_events)
         ))
         # TODO: bound the number we store.
         encoded = json.dumps({
-            "events": [ev.encode() for ev in self._completed_events],
+            "events": [ev.to_json() for ev in self._completed_events],
```
Comment (noahdesu) [inline src/pybind/mgr/progress/module.py]: The JSON encoder for custom objects (e.g. Event in this case) works out pretty well to simplify things. Here is an example https://github.com/ceph/ceph/pull/23497/files#diff-6cb378091f67c72c18f8b31a5fe5a82fR13

File: src/pybind/mgr/progress/module.py [hunk 10]
```diff
@@ -218,6 +235,9 @@ class Module(MgrModule):
         {"cmd": "progress",
          "desc": "Show progress of recovery operations",
          "perm": "r"},
+        {"cmd": "progress json",
+         "desc": "Show machine readable progress information",
+         "perm": "r"},
```
[the remark this answers is not in the document]
Comment (jcsp) [reply src/pybind/mgr/progress/module.py]: You're right -- changed.

**Comment by jcsp** [comment]
Tested here: http://pulpito.ceph.com/jspray-2018-09-11_12:59:10-rados:mgr-wip-jcsp-testing-20180911-distro-basic-smithi/

Will do a follow up to clean up the JSON encoding code + see if we can handle OMAP pool recovery a bit more nicely.

**Pull request merged**